File: Interview Preparation Kit/Search/Pairs.py
#!/bin/python3
# https://www.hackerrank.com/challenges/pairs/problem
import math
import os
import random
import re
import sys
from bisect import bisect

# Complete the pairs function below.
def pairs(k, arr):
    arr.sort()
    result = 0

    edge = len(arr) - 1
    for i in range(edge):
        key = arr[i] + k
        if arr[bisect(arr, key) - 1] == key:
            result += 1

    return result


if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    nk = input().split()

    n = int(nk[0])

    k = int(nk[1])

    arr = list(map(int, input().rstrip().split()))

    result = pairs(k, arr)

    fptr.write(str(result) + '\n')

    fptr.close()

File: Interview Preparation Kit/Search/Pairs.java
// https://www.hackerrank.com/challenges/pairs/problem
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class Solution {

    static int binarysearch(int[] arr, int left, int right, int val) {
        int mid;

        while (left <= right) {
            mid = (left + right) / 2;
            if (arr[mid] == val) {
                return mid;
            }
            else if (arr[mid] > val) {
                right = mid - 1;
            }
            else {
                left = mid + 1;
            }
        }

        return -1;
    }

    // Complete the pairs function below.
    static int pairs(int k, int[] arr) {
        Arrays.sort(arr);
        int result = 0;

        int edge = arr.length - 1;
        for (int i = 0; i < edge; i++) {
            if (binarysearch(arr, 0, edge, arr[i] + k) >= 0) {
                result++;
            }
        }

        return result;
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] nk = scanner.nextLine().split(" ");

        int n = Integer.parseInt(nk[0]);

        int k = Integer.parseInt(nk[1]);

        int[] arr = new int[n];

        String[] arrItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < n; i++) {
            int arrItem = Integer.parseInt(arrItems[i]);
            arr[i] = arrItem;
        }

        int result = pairs(k, arr);

        bufferedWriter.write(String.valueOf(result));
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}
